designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -66,83 +66,164 @@ class DeploymentServiceImpl(
+      ctx <- prepareCommandContextWithAction[CanonicalProcess](
+        processId,
         actionType,
-        d => Some(d.processVersionId),
-        d => Some(d.processingType)
-      ).flatMap { case (processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessIngType) =>
-        validateBeforeDeploy(processDetails, actionId).transformWith {
-          case Failure(ex) =>
-            dbioRunner.runInTransaction(actionRepository.removeAction(actionId)).transform(_ => Failure(ex))
-          case Success(validationResult) =>
-            // we notify of deployment finish/fail only if initial validation succeeded
-            val deploymentFuture = runDeploymentActionWithNotifications(
-              actionType,
-              actionId,
-              processIdWithName,
-              versionOnWhichActionIsDone,
-              validatedComment,
-              buildInfoProcessIngType
-            ) {
-              dispatcher
-                .deploymentManagerUnsafe(processDetails.processingType)
-                .processCommand(
-                  RunDeploymentCommand(
-                    validationResult.processVersion,
-                    validationResult.deploymentData,
-                    validationResult.resolvedScenario,
-                    savepointPath
-                  )
-                )
-            }
-            Future.successful(deploymentFuture)
-        }
+        getCurrentVersion,
+        p => Some(p.processingType)
+      )
+      // 2. command specific section
+      deployedScenarioData <- prepareDeployedScenarioData(ctx.processDetails, ctx.actionId)
+      command = RunDeploymentCommand(
+        deployedScenarioData.processVersion,
+        deployedScenarioData.deploymentData,
+        deployedScenarioData.resolvedScenario,
+        savepointPath
+      )
+      // TODO: move validateBeforeDeploy before creating an action
+      result <- validateBeforeDeploy(ctx.processDetails, deployedScenarioData).transformWith {
+        case Failure(ex) =>
+          removeInvalidAction(ctx.actionId).transform(_ => Failure(ex))
+        case Success(_) =>
+          // we notify of deployment finish/fail only if initial validation succeeded
+          val deploymentFuture = runActionAndHandleResults(
+            actionType,
+            validatedComment,
+            ctx
+          ) {
+            dispatcher
+              .deploymentManagerUnsafe(ctx.processDetails.processingType)
+              .processCommand(command)
+          }
+          Future.successful(deploymentFuture)
       }
+    } yield result
+  }
+
+  /**
+    * Common validations and operations for a command execution.
+    * @return gathered data for further command execution
+    */
+  private def prepareCommandContextWithAction[PS: ScenarioShapeFetchStrategy](
+      processIdWithName: ProcessIdWithName,
+      actionType: ProcessActionType,
+      getScenarioWithVersionOnWhichActionIsDone: ScenarioWithDetailsEntity[PS] => DB[
```

Comment:
It is important to name somehow what is passed in this function maybe we should rename this to `extract...FromLatestScenarioDetails` ?